src/NuGet.Core/NuGet.Common/Migrations/MigrationRunner.cs

```diff
@@ -32,31 +32,39 @@ public static void Run()
                 // so use a global mutex and then check if someone else already did the work.
                 using (var mutex = new Mutex(false, "NuGet-Migrations"))
                 {
-                    bool signal = mutex.WaitOne(TimeSpan.FromMinutes(1), false);
-                    if (signal && !File.Exists(expectedMigrationFilename))
+                    try
                     {
-                        // Only run migrations that have not already been run
-                        int highestMigrationRun = GetHighestMigrationRun(migrationsDirectory);
-                        for (int i = highestMigrationRun + 1; i < Migrations.Count; i++)
+                        if (mutex.WaitOne(TimeSpan.FromMinutes(1), false))
                         {
-                            try
+                            if (!File.Exists(expectedMigrationFilename))
                             {
-                                Migrations[i]();
-                                // Create file for every migration run, so that if an older version of NuGet is run, it doesn't try to run
-                                // migrations again.
-                                string migrationFile = Path.Combine(migrationsDirectory, (i + 1).ToString(CultureInfo.InvariantCulture));
-                                File.WriteAllText(migrationFile, string.Empty);
+                                // Only run migrations that have not already been run
+                                int highestMigrationRun = GetHighestMigrationRun(migrationsDirectory);
+                                for (int i = highestMigrationRun + 1; i < Migrations.Count; i++)
+                                {
+                                    try
+                                    {
+                                        Migrations[i]();
+                                        // Create file for every migration run, so that if an older version of NuGet is run, it doesn't try to run
+                                        // migrations again.
+                                        string migrationFile = Path.Combine(migrationsDirectory, (i + 1).ToString(CultureInfo.InvariantCulture));
+                                        File.WriteAllText(migrationFile, string.Empty);
+                                    }
+                                    catch { }
+                                }
                             }
-                            catch { }
+                            mutex.ReleaseMutex();
                         }
-
-                        mutex.ReleaseMutex();
+                    }
+                    catch (AbandonedMutexException ex)
```

Comment:
Shouldn't the `catch (AbandonedMutexExcention)` be before the `if (!File.Exists...)` so that the migration actually runs?  My understanding of the docs is that `AbandonedMutexException` means that the current process successfully obtained the mutex, just that the other process failed to release it cleanly.  If the other process (or thread?) was killed, then migrations might not have completed, so the current thread should probably ensure that all migrations completed successfully.

Reply:
Make sense Andy. I pushed a new commit to ensure migrations at least once in the event of `AbandonedMutexException`. Please review and let me know the feedback.